internal/pkg/git/git.go

```diff
@@ -12,40 +12,93 @@ import (
 	"github.com/keboola/keboola-as-code/internal/pkg/filesystem"
 	"github.com/keboola/keboola-as-code/internal/pkg/filesystem/aferofs"
 	"github.com/keboola/keboola-as-code/internal/pkg/log"
+	"github.com/keboola/keboola-as-code/internal/pkg/model"
+	"github.com/keboola/keboola-as-code/internal/pkg/template"
+	"github.com/keboola/keboola-as-code/internal/pkg/template/repository/manifest"
 )
 
 func Available() bool {
 	_, err := exec.LookPath("git")
 	return err == nil
 }
 
-func CheckoutTemplateRepository(url string, ref string, logger log.Logger) (filesystem.Fs, error) {
+func CheckoutTemplateRepository(ref model.TemplateRef, logger log.Logger) (filesystem.Fs, error) {
 	if !Available() {
 		return nil, fmt.Errorf("git command is not available, if you want to use templates from a git repository you have to install it first")
 	}
 
+	// Create a temp dir
 	dir, err := ioutil.TempDir("", "keboola-as-code-templates-")
 	if err != nil {
 		return nil, err
 	}
+	defer func(path string) {
+		_ = os.RemoveAll(path) // nolint: forbidigo
+	}(dir)
 
-	err, stdErr, exitCode := runGitCommand(logger, dir, []string{"clone", "--depth", "1", "--no-checkout", "-q", url, dir})
+	// Clone the repository
+	err, stdErr, exitCode := runGitCommand(logger, dir, []string{"clone", "--branch", ref.Repository().Ref, "--depth=1", "--no-checkout", "--sparse", "--filter=blob:none", "-q", ref.Repository().Url, dir})
 	if err != nil {
 		if exitCode == 128 {
-			return nil, fmt.Errorf(`git repository not found on url "%s"`, url)
+			if strings.Contains(stdErr, fmt.Sprintf("Remote branch %s not found", ref.Repository().Ref)) {
+				return nil, fmt.Errorf(`reference "%s" not found in the templates git repository "%s"`, ref.Repository().Ref, ref.Repository().Url)
+			}
+			return nil, fmt.Errorf(`templates git repository not found on url "%s"`, ref.Repository().Url)
 		}
 		return nil, fmt.Errorf(stdErr)
 	}
 
-	err, stdErr, exitCode = runGitCommand(logger, dir, []string{"checkout", ref})
+	// Checkout repository.json
+	err, stdErr, _ = runGitCommand(logger, dir, []string{"sparse-checkout", "add", "/.keboola/repository.json"})
+	if err != nil {
+		return nil, fmt.Errorf(stdErr)
+	}
+	err, stdErr, _ = runGitCommand(logger, dir, []string{"checkout"})
+	if err != nil {
+		return nil, fmt.Errorf(stdErr)
+	}
+
+	// Create FS from the cloned repository
+	localFs, err := aferofs.NewLocalFs(logger, dir, "")
+	if err != nil {
+		return nil, err
+	}
+
+	// Load the repository manifest
+	m, err := manifest.Load(localFs)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get version record
+	version := ref.Version()
+	versionRecord, err := m.GetVersion(ref.TemplateId(), version)
+	if err != nil {
+		// version or template not found
+		return nil, fmt.Errorf(`template "%s" in version "%s" not found in the templates git repository "%s"`, ref.TemplateId(), version.String(), ref.Repository().Url)
```

Comment:
myslim, ze `m.GetVersion` vracia presnejsiu hlasku, tj. ci sa nenasla sablona, alebo verzia.

https://github.com/keboola/keboola-as-code/blob/e2a52384cbdc31213e9ada2cd945975f807399cc/internal/pkg/template/repository/manifest/manifest_test.go#L229-L237

rozmyslam, ci je to potrebne este nejak doplnovat, mozes tam pridat url:
```
fmt.Errorf(`%w, searched in the git repository "%s", ref "%s"`, err, ...)
```

```
template "foo" found, but version "v2" is missing, searched in the git repository "https://", ref "main".
```

Alebo este lepsie pouzit
```go
	if err != nil {
		e := utils.NewMultiError()
		e.Append(fmt.Errorf(`searched in the git repository "%s"`, ref.Repository().Url))
		e.Append(fmt.Errorf(`reference "%s"`, ref.Repository().Ref))
		return nil, utils.PrefixError(err.Error(), e)
	}
```


```
template "foo" found, but version "v2" is missing:
  - searched in the git repository "https://github.com/foo/bar"
  - reference "main"
```

------------

viem, ze je to detail, ... ale napr. niekto bude testovat sablony oproti `dev` branch, a neuvedomi si, ze to nezmenil v `manifest.json`, a spusta to oproti `main`. A bude nam pisat na support :) tak preto chcem, aby ta chyba bola jednoznacna a prehladna.

Reply:
jasně, jsem pro 👍